providers/base/bin/camera_test.py

```diff
@@ -282,89 +370,204 @@ def _setup(self, sink=None):
             "-r",
             "%dx%d" % (width, height),
             filename,
         ]
-        use_camerabin = False
         if pixelformat:
             if "MJPG" == pixelformat:  # special tweak for fswebcam
                 pixelformat = "MJPEG"
             command.extend(["-p", pixelformat])
-
         try:
             check_call(command, stdout=open(os.devnull, "w"), stderr=STDOUT)
             if os.path.getsize(filename) == 0:
-                use_camerabin = True
+                return False
         except (CalledProcessError, OSError):
-            use_camerabin = True
-        if use_camerabin:
-            self._setup(sink="fakesink")
-            GLib.timeout_add_seconds(3, self._take_photo, filename)
-            GLib.timeout_add_seconds(4, self._stop)
-            Gtk.main()
-        if not quiet:
-            stage = Clutter.Stage()
-            stage.set_title("Camera still picture test")
-            stage.set_size(width, height)
-            stage.connect("destroy", self._on_destroy)
-            Clutter.threads_add_timeout(0, 10000, self._on_destroy, None, None)
-            still_texture = Clutter.Texture.new_from_file(filename)
-            stage.add_actor(still_texture)
-            stage.show()
-            Clutter.main()
+            return False
+        return True
+
+    def _capture_image_gstreamer(self, filename, width, height, pixelformat):
+        """
+        Setup the gstreamer pipeline to capture an image.
+        This pipeline consists of the following elements:
+        - v4l2src: Capture video from a V4L2 device
+        - caps: Filter the video stream to the desired resolution
+        - bayer2rgb (only for bayer format): Convert the bayer format to RGB
+        - jpegenc: Encode the video stream to a JPEG image
+        - filesink: Save the JPEG image to a file
+        """
+
+        pipeline = self.Gst.Pipeline.new("pipeline")
+
+        # Add source
+        source = self.Gst.ElementFactory.make("v4l2src", "video-source")
+        source.set_property("device", self.device)
+        source.set_property("num-buffers", 10)
+        pipeline.add(source)
+
+        # Add caps
+        caps = self.Gst.ElementFactory.make("capsfilter", "caps")
+        if pixelformat == "RG10":
+            caps.set_property(
+                "caps",
+                self.Gst.Caps.from_string(
+                    "video/x-bayer, format=gbrg, width={},height={}".format(
+                        width, height
+                    )
+                ),
+            )
+            # Add bayer2rgb
+            bayer2rgb = self.Gst.ElementFactory.make("bayer2rgb", "bayer2rgb")
+
+            pipeline.add(caps)
+            pipeline.add(bayer2rgb)
+        else:
+            caps.set_property(
+                "caps",
+                self.Gst.Caps.from_string(
+                    "video/x-raw,width={},height={}".format(width, height)
+                ),
+            )
+            pipeline.add(caps)
```

Comment:
I don't know exactly if there may be more exceptions but we could create a rgb_capture so that we spare an if followup down below and it is clearer what this is doing on frist glance
```suggestion
        if pixelformat == "RG10":
            caps.set_property(
                "caps",
                self.Gst.Caps.from_string(
                    "video/x-bayer, format=gbrg, width={},height={}".format(
                        width, height
                    )
                ),
            )
            # Add bayer2rgb
            bayer2rgb = self.Gst.ElementFactory.make("bayer2rgb", "bayer2rgb")

            pipeline.add(caps)
            pipeline.add(bayer2rgb)
            caps.link(bayer2rgb)
            # rgb_capture is the node that always outputs rgb
            rgb_capture = bayer2rgb
        else:
            caps.set_property(
                "caps",
                self.Gst.Caps.from_string(
                    "video/x-raw,width={},height={}".format(width, height)
                ),
            )
            pipeline.add(caps)
            rgb_capture = caps
```

> Note: I don't know if you can actually link here or you must do them in order, give it a look

Reply:
That would work more or less. You can link them in a different order, but you also have to link the source to the caps element, so you have to define the starting and final element of the "rgb_capture'' pipeline. 
That could be done defining a bin that contains it, but it's a bit overkill for such a simple case, since you also have to define "ghost pads" for the source and the sink